functions/max_active_instances.pp

```diff
@@ -3,22 +3,57 @@
 #
 # Parameters are not to be used but are present as an assist to testing
 #
-# @param memory_limited_instances
-# @param processor_count
+# @param server_type
 #
 # @return [Integer]
-#   The maximum number of instances that should be active on the ``puppetserver``
+#   The maximum number of JRuby instances that should be active on the ``puppetserver``
 #
 function pupmod::max_active_instances (
-  Integer[0] $memory_limited_instances = Integer(pick(fact('memorysize_mb'), 1) / 512),
-  Integer[1] $processor_count          = pick(fact('processors.count'), 1)
-){
+  Enum['monolithic', 'primary', 'compile'] $server_type = 'monolithic'
+) {
+  $processor_count = pick(fact('processors.count'), 1)
 
-  if ($memory_limited_instances < 2) or ($processor_count < 3) {
-    $max_active_instances = 1
+  # Figure out memory limited instances up front
+  if $processor_count < 8 {
+    $per_instance_mem = 512
+  } elsif $processor_count < 16 {
+    $per_instance_mem = 768
+  } else {
+    $per_instance_mem = 1024
   }
-  else {
-    $max_active_instances = min($memory_limited_instances, $processor_count) - 1
+
+  $_floor_mem_instances = floor(fact('memorysize_mb') / $per_instance_mem)
+  $memory_limited_instances = $_floor_mem_instances ? {
+    0       => 1,
+    default => $_floor_mem_instances,
+  }
+
+  if $server_type == 'monolithic' {
+    if $processor_count < 8 {
+      $_ratio = (1 / 2)
+    } elsif $processor_count < 16 {
+      $_ratio = (5 / 8)
+    } else {
+      $_ratio = (11 / 16)
+    }
+
+    $cpu_limited_instances = max(floor($processor_count * $_ratio), 1)
+
+    $max_active_instances = min($cpu_limited_instances, $memory_limited_instances)
+  } elsif $server_type == 'primary' {
+    # Calculations are simpler here
+    if $processor_count < 4 {
```

Comment:
Not quite sure why this is 4.

Maybe:
* puppetserver => 1
* puppetdb => 1
* puppet runs => 1
* other => 1

This makes sense overall but I'm just not sure where the 4 comes from.

Reply:
These recommendations come straight out of the tuning guide for PE, and out of the puppet infrastructure tune tool (minus some calculations based on catalog times) for when using primary server and compile server setups. But yes, essentially this gives the other primary processes (such as PuppetDB, Postgres, etc.) cores to utilize and since there should only be the minimal set of compile servers themselves checking into the primary server, the reduced load is still covered.